flink-streaming-connectors/flink-connector-activemq/src/main/java/org/apache/flink/streaming/connectors/activemq/AMQSource.java

```diff
@@ -0,0 +1,258 @@
+				LOG.error("Received ActiveMQ exception", e);
+			}
+		});
+
+		RuntimeContext runtimeContext = getRuntimeContext();
+		int acknowledgeType;
+		if (runtimeContext instanceof StreamingRuntimeContext
+			&& ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled()) {
+			autoAck = false;
+			acknowledgeType = ActiveMQSession.INDIVIDUAL_ACKNOWLEDGE;
+		} else {
+			autoAck = true;
+			acknowledgeType = ActiveMQSession.AUTO_ACKNOWLEDGE;
+		}
+		// Create a Session
+		session = connection.createSession(false, acknowledgeType);
+
+		// Create the destination (Topic or Queue)
+		Destination destination = session.createQueue(queueName);
+
+		// Create a MessageConsumer from the Session to the Topic or
+		// Queue
+		consumer = session.createConsumer(destination);
+		runningChecker.setIsRunning(true);
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		RuntimeException exception = null;
+		try {
+			consumer.close();
+		} catch (JMSException e) {
+			if (logFailuresOnly) {
+				LOG.error("Failed to close ActiveMQ session", e);
+			} else {
+				exception = new RuntimeException("Failed to close ActiveMQ consumer", e);
+			}
+		}
+		try {
+			session.close();
+		} catch (JMSException e) {
+			if (logFailuresOnly) {
+				LOG.error("Failed to close ActiveMQ session", e);
+			} else {
+				exception = exception == null	? new RuntimeException("Failed to close ActiveMQ session", e)
+												: exception;
+			}
+
+		}
+		try {
+			connection.close();
+		} catch (JMSException e) {
+			if (logFailuresOnly) {
+				LOG.error("Failed to close ActiveMQ session", e);
+			} else {
+				exception = exception == null	? new RuntimeException("Failed to close ActiveMQ connection", e)
+												: exception;
+			}
+		}
+
+		if (exception != null) {
+			throw exception;
+		}
+	}
+
+	@Override
+	protected void acknowledgeIDs(long checkpointId, List<String> UIds) {
+		try {
+			for (String messageId : UIds) {
+				Message unacknowledgedMessage = unaknowledgedMessages.get(messageId);
+				if (unacknowledgedMessage != null) {
+					unacknowledgedMessage.acknowledge();
+					unaknowledgedMessages.remove(messageId);
+				} else {
+					LOG.warn("Tried to acknowledge unknown ActiveMQ message id: {}", messageId);
+				}
+			}
+		} catch (JMSException e) {
+			if (logFailuresOnly) {
+				LOG.error("Failed to acknowledge ActiveMQ message");
+			} else {
+				throw new RuntimeException("Failed to acknowledge ActiveMQ message");
+			}
+		}
+	}
+
+	@Override
+	public void run(SourceContext<OUT> ctx) throws Exception {
+		while (runningChecker.isRunning()) {
+			Message message = consumer.receive(1000);
+			synchronized (ctx.getCheckpointLock()) {
+				if (message instanceof BytesMessage) {
+					BytesMessage bytesMessage = (BytesMessage) message;
+					byte[] bytes = new byte[(int) bytesMessage.getBodyLength()];
+					bytesMessage.readBytes(bytes);
+					OUT value = deserializationSchema.deserialize(bytes);
+					ctx.collect(value);
```

Comment:
I think you need to hold the checkpoint lock only for the collect() call and the if() below.